File: frontend/packages/ui/src/utils/chatCompletionRecovery.ts
/**
 * Browser-compatible chat completion recovery cryptography.
 *
 * Clients deterministically derive an X25519 recovery keypair from the local
 * chat key. Only the public key and authenticated sealed envelopes cross the
 * server boundary; private and plaintext material remain client-side.
 */

import nacl from "tweetnacl";

const PROTOCOL_VERSION = 1;
const MAX_PAYLOAD_BYTES = 16 * 1024 * 1024;
const KEY_BYTES = 32;
const NONCE_BYTES = 12;
const AAD_PREFIX = new TextEncoder().encode("OMCR1");
const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

export interface ChatCompletionRecoveryIdentity {
  ownerId: string;
  chatId: string;
  turnId: string;
  jobId: string;
  assistantMessageId: string;
  keyVersion: number;
}

export interface ChatCompletionRecoveryEnvelope {
  v: number;
  epk: string;
  nonce: string;
  ciphertext: string;
}

function toArrayBuffer(input: Uint8Array): ArrayBuffer {
  const output = new ArrayBuffer(input.byteLength);
  new Uint8Array(output).set(input);
  return output;
}

function encodeBase64Url(input: Uint8Array): string {
  let binary = "";
  const chunkSize = 0x8000;
  for (let offset = 0; offset < input.length; offset += chunkSize) {
    const chunk = input.subarray(offset, offset + chunkSize);
    for (let index = 0; index < chunk.length; index += 1) {
      binary += String.fromCharCode(chunk[index]);
    }
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function decodeBase64Url(input: string, field: string, expectedLength?: number): Uint8Array {
  if (!input || input.includes("=") || !/^[A-Za-z0-9_-]+$/.test(input)) {
    throw new Error(`${field} must be non-empty unpadded base64url`);
  }
  const standard = input.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat(-input.length & 3);
  let binary: string;
  try {
    binary = atob(standard);
  } catch {
    throw new Error(`${field} must be canonical base64url`);
  }
  const decoded = Uint8Array.from(binary, (character) => character.charCodeAt(0));
  if (encodeBase64Url(decoded) !== input) {
    throw new Error(`${field} must be canonical base64url`);
  }
  if (expectedLength !== undefined && decoded.length !== expectedLength) {
    throw new Error(`${field} must decode to ${expectedLength} bytes`);
  }
  return decoded;
}

function uint32(value: number, field: string): Uint8Array {
  if (!Number.isInteger(value) || value < 0 || value > 0xffffffff) {
    throw new Error(`${field} must be an unsigned 32-bit integer`);
  }
  const encoded = new Uint8Array(4);
  new DataView(encoded.buffer).setUint32(0, value, false);
  return encoded;
}

function canonicalUuid(value: string, field: string): Uint8Array {
  if (!UUID_PATTERN.test(value)) {
    throw new Error(`${field} must be a canonical lowercase UUID`);
  }
  return new TextEncoder().encode(value);
}

function concatBytes(...values: Uint8Array[]): Uint8Array {
  const output = new Uint8Array(values.reduce((total, value) => total + value.length, 0));
  let offset = 0;
  for (const value of values) {
    output.set(value, offset);
    offset += value.length;
  }
  return output;
}

function lengthPrefix(value: Uint8Array): Uint8Array {
  return concatBytes(uint32(value.length, "length"), value);
}

async function sha256(input: Uint8Array): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.digest("SHA-256", toArrayBuffer(input)));
}

async function hkdfSha256(input: Uint8Array, salt: Uint8Array, info: Uint8Array): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey("raw", toArrayBuffer(input), "HKDF", false, ["deriveBits"]);
  return new Uint8Array(await crypto.subtle.deriveBits(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: toArrayBuffer(salt),
      info: toArrayBuffer(info),
    },
    key,
    256,
  ));
}

export async function deriveChatCompletionRecoveryKeypair(
  chatKey: string,
  chatId: string,
  keyVersion: number,
): Promise<{ privateKey: string; publicKey: string }> {
  const rawChatKey = decodeBase64Url(chatKey, "chat_key", KEY_BYTES);
  const salt = await sha256(new TextEncoder().encode("openmates:chat-recovery:v1"));
  const info = concatBytes(lengthPrefix(canonicalUuid(chatId, "chat_id")), uint32(keyVersion, "key_version"));
  const privateKey = await hkdfSha256(rawChatKey, salt, info);
  return {
    privateKey: encodeBase64Url(privateKey),
    publicKey: encodeBase64Url(nacl.scalarMult.base(privateKey)),
  };
}

export function buildRecoveryAssociatedData(values: {
  owner_id: string;
  chat_id: string;
  turn_id: string;
  job_id: string;
  assistant_message_id: string;
  key_version: number;
}): string {
  return encodeBase64Url(concatBytes(
    AAD_PREFIX,
    lengthPrefix(canonicalUuid(values.owner_id, "owner_id")),
    lengthPrefix(canonicalUuid(values.chat_id, "chat_id")),
    lengthPrefix(canonicalUuid(values.turn_id, "turn_id")),
    lengthPrefix(canonicalUuid(values.job_id, "job_id")),
    lengthPrefix(canonicalUuid(values.assistant_message_id, "assistant_message_id")),
    uint32(values.key_version, "key_version"),
  ));
}

function associatedData(identity: ChatCompletionRecoveryIdentity): Uint8Array {
  return decodeBase64Url(buildRecoveryAssociatedData({
    owner_id: identity.ownerId,
    chat_id: identity.chatId,
    turn_id: identity.turnId,
    job_id: identity.jobId,
    assistant_message_id: identity.assistantMessageId,
    key_version: identity.keyVersion,
  }), "associated_data");
}

async function envelopeKey(sharedSecret: Uint8Array, aad: Uint8Array): Promise<Uint8Array> {
  if (sharedSecret.every((value) => value === 0)) {
    throw new Error("X25519 shared secret must not be all zero");
  }
  const salt = await sha256(new TextEncoder().encode("openmates:chat-recovery-envelope:v1"));
  return hkdfSha256(sharedSecret, salt, await sha256(aad));
}

export async function sealChatCompletionRecoveryPayload(
  plaintext: Uint8Array,
  options: ChatCompletionRecoveryIdentity & {
    recoveryPublicKey: string;
    ephemeralPrivateKey?: string;
    nonce?: string;
  },
): Promise<ChatCompletionRecoveryEnvelope> {
  if (plaintext.length > MAX_PAYLOAD_BYTES) {
    throw new Error(`plaintext must be no larger than ${MAX_PAYLOAD_BYTES}`);
  }
  const recoveryPublicKey = decodeBase64Url(options.recoveryPublicKey, "recovery_public_key", KEY_BYTES);
  const ephemeralPrivateKey = options.ephemeralPrivateKey
    ? decodeBase64Url(options.ephemeralPrivateKey, "ephemeral_private_key", KEY_BYTES)
    : crypto.getRandomValues(new Uint8Array(KEY_BYTES));
  const nonce = options.nonce
    ? decodeBase64Url(options.nonce, "nonce", NONCE_BYTES)
    : crypto.getRandomValues(new Uint8Array(NONCE_BYTES));
  const ephemeralPublicKey = nacl.scalarMult.base(ephemeralPrivateKey);
  const aad = associatedData(options);
  const key = await crypto.subtle.importKey(
    "raw",
    toArrayBuffer(await envelopeKey(nacl.scalarMult(ephemeralPrivateKey, recoveryPublicKey), aad)),
    { name: "AES-GCM" },
    false,
    ["encrypt"],
  );
  const ciphertext = new Uint8Array(await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: toArrayBuffer(nonce), additionalData: toArrayBuffer(aad) },
    key,
    toArrayBuffer(plaintext),
  ));
  return {
    v: PROTOCOL_VERSION,
    epk: encodeBase64Url(ephemeralPublicKey),
    nonce: encodeBase64Url(nonce),
    ciphertext: encodeBase64Url(ciphertext),
  };
}

export async function openChatCompletionRecoveryEnvelope(
  envelope: ChatCompletionRecoveryEnvelope,
  options: ChatCompletionRecoveryIdentity & { recoveryPrivateKey: string },
): Promise<Uint8Array> {
  if (Object.keys(envelope).sort().join(",") !== "ciphertext,epk,nonce,v" || envelope.v !== PROTOCOL_VERSION) {
    throw new Error("invalid recovery envelope fields or version");
  }
  const recoveryPrivateKey = decodeBase64Url(options.recoveryPrivateKey, "recovery_private_key", KEY_BYTES);
  const ephemeralPublicKey = decodeBase64Url(envelope.epk, "epk", KEY_BYTES);
  const nonce = decodeBase64Url(envelope.nonce, "nonce", NONCE_BYTES);
  const ciphertext = decodeBase64Url(envelope.ciphertext, "ciphertext");
  if (ciphertext.length < 16 || ciphertext.length - 16 > MAX_PAYLOAD_BYTES) {
    throw new Error("ciphertext payload size is invalid");
  }
  const aad = associatedData(options);
  const key = await crypto.subtle.importKey(
    "raw",
    toArrayBuffer(await envelopeKey(nacl.scalarMult(recoveryPrivateKey, ephemeralPublicKey), aad)),
    { name: "AES-GCM" },
    false,
    ["decrypt"],
  );
  return new Uint8Array(await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: toArrayBuffer(nonce), additionalData: toArrayBuffer(aad) },
    key,
    toArrayBuffer(ciphertext),
  ));
}

File: backend/shared/python_utils/chat_completion_recovery.py
"""
Cryptographic primitives for sealed cross-device chat completion recovery.

This module implements the versioned byte contract shared by backend and all
clients. The backend seals with public recovery keys only; raw chat keys and
derived recovery private keys remain client-side.
"""

from __future__ import annotations

import base64
import hashlib
import os
import struct
import uuid
from collections.abc import Mapping
from typing import Any

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import x25519
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF


PROTOCOL_VERSION = 1
KEY_LENGTH = 32
NONCE_LENGTH = 12
MAX_PAYLOAD_BYTES = 16 * 1024 * 1024
RECOVERY_KEY_SALT = hashlib.sha256(b"openmates:chat-recovery:v1").digest()
ENVELOPE_KEY_SALT = hashlib.sha256(b"openmates:chat-recovery-envelope:v1").digest()
ENVELOPE_FIELDS = {"v", "epk", "nonce", "ciphertext"}


def _encode(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _decode(value: str, field: str) -> bytes:
    if not isinstance(value, str) or not value or "=" in value:
        raise ValueError(f"{field} must be non-empty unpadded base64url")
    try:
        decoded = base64.b64decode(value + "=" * (-len(value) % 4), altchars=b"-_", validate=True)
    except (ValueError, TypeError) as exc:
        raise ValueError(f"{field} must be valid unpadded base64url") from exc
    if _encode(decoded) != value:
        raise ValueError(f"{field} must use canonical unpadded base64url")
    return decoded


def _canonical_uuid(value: str, field: str) -> str:
    try:
        canonical = str(uuid.UUID(value))
    except (ValueError, AttributeError) as exc:
        raise ValueError(f"{field} must be a UUID") from exc
    if canonical != value:
        raise ValueError(f"{field} must use canonical lowercase UUID encoding")
    return value


def _key_version(value: int) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or not 1 <= value <= 0xFFFFFFFF:
        raise ValueError("key_version must be an unsigned 32-bit integer greater than zero")
    return value


def _length_prefixed(value: str) -> bytes:
    encoded = value.encode("utf-8")
    return struct.pack(">I", len(encoded)) + encoded


def _associated_data(
    *,
    owner_id: str,
    chat_id: str,
    turn_id: str,
    job_id: str,
    assistant_message_id: str,
    key_version: int,
) -> bytes:
    identifiers = (
        _canonical_uuid(owner_id, "owner_id"),
        _canonical_uuid(chat_id, "chat_id"),
        _canonical_uuid(turn_id, "turn_id"),
        _canonical_uuid(job_id, "job_id"),
        _canonical_uuid(assistant_message_id, "assistant_message_id"),
    )
    return b"OMCR1" + b"".join(_length_prefixed(value) for value in identifiers) + struct.pack(">I", _key_version(key_version))


def build_recovery_associated_data(values: Mapping[str, Any]) -> str:
    return _encode(
        _associated_data(
            owner_id=values["owner_id"],
            chat_id=values["chat_id"],
            turn_id=values["turn_id"],
            job_id=values["job_id"],
            assistant_message_id=values["assistant_message_id"],
            key_version=values["key_version"],
        )
    )


def derive_recovery_keypair(chat_key: str, chat_id: str, key_version: int) -> tuple[str, str]:
    chat_key_bytes = _decode(chat_key, "chat_key")
    if len(chat_key_bytes) != KEY_LENGTH:
        raise ValueError("chat_key must contain exactly 32 bytes")
    private_bytes = HKDF(
        algorithm=hashes.SHA256(),
        length=KEY_LENGTH,
        salt=RECOVERY_KEY_SALT,
        info=_length_prefixed(_canonical_uuid(chat_id, "chat_id")) + struct.pack(">I", _key_version(key_version)),
    ).derive(chat_key_bytes)
    private_key = x25519.X25519PrivateKey.from_private_bytes(private_bytes)
    public_bytes = private_key.public_key().public_bytes(
        serialization.Encoding.Raw,
        serialization.PublicFormat.Raw,
    )
    return _encode(private_bytes), _encode(public_bytes)


def _envelope_key(private_key: x25519.X25519PrivateKey, public_key: bytes, associated_data: bytes) -> bytes:
    if len(public_key) != KEY_LENGTH:
        raise ValueError("X25519 public key must contain exactly 32 bytes")
    shared_secret = private_key.exchange(x25519.X25519PublicKey.from_public_bytes(public_key))
    return HKDF(
        algorithm=hashes.SHA256(),
        length=KEY_LENGTH,
        salt=ENVELOPE_KEY_SALT,
        info=hashlib.sha256(associated_data).digest(),
    ).derive(shared_secret)


def seal_recovery_payload(
    payload: bytes,
    *,
    recovery_public_key: str,
    owner_id: str,
    chat_id: str,
    turn_id: str,
    job_id: str,
    assistant_message_id: str,
    key_version: int,
    ephemeral_private_key: str | None = None,
    nonce: str | None = None,
) -> dict[str, str | int]:
    if len(payload) > MAX_PAYLOAD_BYTES:
        raise ValueError("recovery payload exceeds 16 MiB")
    ephemeral_bytes = _decode(ephemeral_private_key, "ephemeral_private_key") if ephemeral_private_key else os.urandom(KEY_LENGTH)
    nonce_bytes = _decode(nonce, "nonce") if nonce else os.urandom(NONCE_LENGTH)
    if len(ephemeral_bytes) != KEY_LENGTH or len(nonce_bytes) != NONCE_LENGTH:
        raise ValueError("invalid ephemeral key or nonce length")
    associated_data = _associated_data(
        owner_id=owner_id,
        chat_id=chat_id,
        turn_id=turn_id,
        job_id=job_id,
        assistant_message_id=assistant_message_id,
        key_version=key_version,
    )
    ephemeral = x25519.X25519PrivateKey.from_private_bytes(ephemeral_bytes)
    key = _envelope_key(ephemeral, _decode(recovery_public_key, "recovery_public_key"), associated_data)
    ephemeral_public = ephemeral.public_key().public_bytes(
        serialization.Encoding.Raw,
        serialization.PublicFormat.Raw,
    )
    return {
        "v": PROTOCOL_VERSION,
        "epk": _encode(ephemeral_public),
        "nonce": _encode(nonce_bytes),
        "ciphertext": _encode(AESGCM(key).encrypt(nonce_bytes, payload, associated_data)),
    }


def open_recovery_envelope(
    envelope: Mapping[str, Any],
    *,
    recovery_private_key: str,
    owner_id: str,
    chat_id: str,
    turn_id: str,
    job_id: str,
    assistant_message_id: str,
    key_version: int,
) -> bytes:
    if set(envelope) != ENVELOPE_FIELDS or envelope.get("v") != PROTOCOL_VERSION:
        raise ValueError("unsupported recovery envelope")
    private_bytes = _decode(recovery_private_key, "recovery_private_key")
    ephemeral_public = _decode(envelope["epk"], "epk")
    nonce = _decode(envelope["nonce"], "nonce")
    ciphertext = _decode(envelope["ciphertext"], "ciphertext")
    if len(private_bytes) != KEY_LENGTH or len(nonce) != NONCE_LENGTH:
        raise ValueError("invalid recovery key or nonce length")
    if len(ciphertext) > MAX_PAYLOAD_BYTES + 16:
        raise ValueError("recovery payload exceeds 16 MiB")
    associated_data = _associated_data(
        owner_id=owner_id,
        chat_id=chat_id,
        turn_id=turn_id,
        job_id=job_id,
        assistant_message_id=assistant_message_id,
        key_version=key_version,
    )
    private_key = x25519.X25519PrivateKey.from_private_bytes(private_bytes)
    key = _envelope_key(private_key, ephemeral_public, associated_data)
    return AESGCM(key).decrypt(nonce, ciphertext, associated_data)
